loc_books/books.py

```diff
@@ -0,0 +1,213 @@
+)
+from tqdm import tqdm
+
+from licensed_pile import logs
+from licensed_pile.licenses import PermissiveLicenses
+from licensed_pile.write import to_dolma
+
+data_path = Path(__file__).resolve().parent / "data"
+
+metadata_exports_path = data_path / "exports/metadata"
+book_downloads_path = data_path / "downloads/books"
+book_exports_path = data_path / "exports/books"
+
+book_downloads_path.mkdir(parents=True, exist_ok=True)
+book_exports_path.mkdir(parents=True, exist_ok=True)
+
+logger = logs.get_logger("loc_books")
+
+
+class LocBooksDownloader:
+    def __init__(self, snapshot):
+        self.snapshot = snapshot
+
+        metadata_path = Path(metadata_exports_path) / f"{snapshot}.csv"
+        self.metadata_file = metadata_path.absolute()
+
+        self.concurrent_rate = RequestRate(10, Duration.SECOND)
+        self.burst_rate = RequestRate(80, Duration.SECOND * 10)
+        self.crawl_rate = RequestRate(400, Duration.MINUTE)
+        self.limiter = Limiter(self.concurrent_rate, self.burst_rate, self.crawl_rate)
+        self.session = LimiterSession(limiter=self.limiter)
+
+        self.progress_bar = tqdm(total=0, position=0, delay=1, desc="Downloading books")
+
+    def download(self):
+        metadata = pd.read_csv(self.metadata_file)
+        metadata_filtered = metadata.dropna()
+        metadata_filtered = metadata_filtered[
+            metadata_filtered["language"] == "english"
+        ]
+        metadata_filtered = metadata_filtered[metadata_filtered["year"] >= 1500]
+
+        text_file_urls = metadata_filtered["text_file_url"]
+        download_urls = self.urls_to_download(text_file_urls)
+
+        logger.info(f"Found {len(text_file_urls)} books in metadata file")
+        logger.info(
+            f"{len(text_file_urls) - len(download_urls)} books already downloaded"
+        )
+        logger.info(f"Downloading {len(download_urls)} books")
+
+        self.progress_bar.total = len(download_urls)
+
+        with mp.Pool(10) as pool:
+            results = pool.imap(functools.partial(self.download_book), download_urls)
+            for result in results:
+                if result:
+                    self.progress_bar.set_description(f"{result['filename']}")
+                    self.progress_bar.update(1)
+                pass
+
+    def urls_to_download(self, text_file_urls):
+        existing_files = set(
+            [path.name for path in Path(book_downloads_path).glob("*.txt")]
+        )
+        download_urls = [
+            text_file_url
+            for text_file_url in text_file_urls
+            if Path(str(furl(text_file_url).path)).name not in existing_files
+        ]
+
+        return download_urls
+
+    def download_book(self, text_file_url):
+        text_file_furl = furl(text_file_url)
+        text_file_name = Path(str(text_file_furl.path)).name
+        text_file_path = os.path.join(book_downloads_path, text_file_name)
+        if not os.path.exists(text_file_path):
+            try:
```

Comment:
Is there a simpler way to structure this? It seems like:
* The first try is the one that catches an error when you fail all your retries
* The inner try is for non-transient errors?

The loop makes things confusing, can you at least comment what all the bits are doing?